Cannot handle enums with numeric fields

### Expected Behavior

When presented with enum with number fields (e.g. `FOO(1)`), it should respect `@JsonValue` and `@JsonCreator` and should compile.

### Actual Behaviour

When running the test, compilation fails with

```
  Caused by: com.fasterxml.jackson.databind.JsonMappingException: Cannot deserialize value of type `java.lang.Number` from String "EdDSA": not a valid number
```

This works fine in micronaut 3.

The error does not really make sense as the enum should be serialized to a number (based on `@JsonValue`) and never be in the string form. 

### Steps To Reproduce

1. Clone https://github.com/gnarroway-bullish/micronaut-openapi-error
2. Run the test and see compilation fail with the above error (it does not even get to running the test)
3. Adjust the HelloController to use COSEAlgorithmIdentifier2 (with 2 suffix). This uses `@JsonProperty` instead
4. Re-run the test and see it pass

Raising this in micronaut repo, as the same enum class serializes and deserializes fine with jackson by itself.

### Environment Information

- OSX
- JDK 21
- Micronaut 4.0.0 -> 4.7.6

### Example Application

https://github.com/gnarroway-bullish/micronaut-openapi-error

### Version

4.7.6

--- a/openapi/src/main/java/io/micronaut/openapi/visitor/GeneratorUtils.java
+++ b/openapi/src/main/java/io/micronaut/openapi/visitor/GeneratorUtils.java
@@ -135,7 +135,7 @@ public final class GeneratorUtils {
         }
     }
 
-    public static void addEnumExtensions(EnumElement enumEl, Schema<?> schema, VisitorContext context) {
+    public static void addEnumExtensions(EnumElement enumEl, Schema<?> schema, boolean cantDeserializeValues, VisitorContext context) {
 
         if (!ConfigUtils.isGeneratorExtensionsEnabled(context) || enumEl == null) {
             return;
@@ -147,41 +147,43 @@ public final class GeneratorUtils {
             schema.setExtensions(extensions);
         }
 
-        String xType = null;
-        String xFormat = null;
+        if (!cantDeserializeValues) {
+            String xType = null;
+            String xFormat = null;
 
-        ClassElement fieldType = findJsonValueType(enumEl, context);
-        if (fieldType != null) {
-            if (fieldType.isPrimitive()) {
-                xType = fieldType.getSimpleName();
-            } else {
-                xType = fieldType.getSimpleName();
-                if (xType.equalsIgnoreCase("byte")) {
-                    xType = null;
-                    xFormat = "int8";
-                } else if (xType.equalsIgnoreCase("short")) {
-                    xType = null;
-                    xFormat = "int16";
-                } else if (xType.equalsIgnoreCase("int")
-                    || xType.equalsIgnoreCase("integer")
-                    || xType.equalsIgnoreCase("long")
-                    || xType.equalsIgnoreCase("float")
-                    || xType.equalsIgnoreCase("double")) {
-                    xType = null;
-                    // because boolean enums generated as Boolean type
-                } else if (xType.equalsIgnoreCase("boolean")) {
-                    xType = null;
-                } else if (xType.equalsIgnoreCase("char") || xType.equalsIgnoreCase("character")) {
-                    xType = "char";
+            ClassElement fieldType = findJsonValueType(enumEl, context);
+            if (fieldType != null) {
+                if (fieldType.isPrimitive()) {
+                    xType = fieldType.getSimpleName();
+                } else {
+                    xType = fieldType.getSimpleName();
+                    if (xType.equalsIgnoreCase("byte")) {
+                        xType = null;
+                        xFormat = "int8";
+                    } else if (xType.equalsIgnoreCase("short")) {
+                        xType = null;
+                        xFormat = "int16";
+                    } else if (xType.equalsIgnoreCase("int")
+                        || xType.equalsIgnoreCase("integer")
+                        || xType.equalsIgnoreCase("long")
+                        || xType.equalsIgnoreCase("float")
+                        || xType.equalsIgnoreCase("double")) {
+                        xType = null;
+                        // because boolean enums generated as Boolean type
+                    } else if (xType.equalsIgnoreCase("boolean")) {
+                        xType = null;
+                    } else if (xType.equalsIgnoreCase("char") || xType.equalsIgnoreCase("character")) {
+                        xType = "char";
+                    }
                 }
             }
-        }
 
-        if (xType != null && !extensions.containsKey(TYPE)) {
-            extensions.put(TYPE, xType);
-        }
-        if (xFormat != null && !extensions.containsKey(FORMAT)) {
-            extensions.put(FORMAT, xFormat);
+            if (xType != null && !extensions.containsKey(TYPE)) {
+                extensions.put(TYPE, xType);
+            }
+            if (xFormat != null && !extensions.containsKey(FORMAT)) {
+                extensions.put(FORMAT, xFormat);
+            }
         }
 
         var enumVarNameList = new ArrayList<String>();

--- a/openapi/src/main/java/io/micronaut/openapi/visitor/SchemaDefinitionUtils.java
+++ b/openapi/src/main/java/io/micronaut/openapi/visitor/SchemaDefinitionUtils.java
@@ -131,6 +131,7 @@ import static io.micronaut.openapi.visitor.ConfigUtils.getInnerClassSeparator;
 import static io.micronaut.openapi.visitor.ConfigUtils.getSchemaDecoration;
 import static io.micronaut.openapi.visitor.ConfigUtils.getSchemaDuplicateResolution;
 import static io.micronaut.openapi.visitor.ConfigUtils.isJsonViewDefaultInclusion;
+import static io.micronaut.openapi.visitor.ContextUtils.info;
 import static io.micronaut.openapi.visitor.ContextUtils.warn;
 import static io.micronaut.openapi.visitor.ConvertUtils.parseJsonString;
 import static io.micronaut.openapi.visitor.ConvertUtils.setDefaultValueObject;
@@ -222,6 +223,7 @@ import static io.micronaut.openapi.visitor.ProtoUtils.normalizeProtobufClassName
 import static io.micronaut.openapi.visitor.ProtoUtils.protobufTypeSchema;
 import static io.micronaut.openapi.visitor.SchemaUtils.TYPE_ARRAY;
 import static io.micronaut.openapi.visitor.SchemaUtils.TYPE_OBJECT;
+import static io.micronaut.openapi.visitor.SchemaUtils.TYPE_STRING;
 import static io.micronaut.openapi.visitor.SchemaUtils.appendSchema;
 import static io.micronaut.openapi.visitor.SchemaUtils.getSchemaByRef;
 import static io.micronaut.openapi.visitor.SchemaUtils.isEmptySchema;
@@ -311,8 +313,7 @@ public final class SchemaDefinitionUtils {
 
         if (type instanceof EnumElement enumEl && isEnum(enumEl)) {
             if (CollectionUtils.isEmpty(schema.getEnum())) {
-                schema.setEnum(getEnumValues(enumEl, schema.getType(), schema.getFormat(), context));
-                addEnumExtensions(enumEl, schema, context);
+                setEnumValues(enumEl, schema, context);
             }
         } else {
             JavadocDescription javadoc = type != null ? Utils.getJavadocParser().parse(type.getDescription()) : null;
@@ -431,8 +432,7 @@ public final class SchemaDefinitionUtils {
                         schema.setType(typeAndFormat.getFirst());
                         schema.setFormat(typeAndFormat.getSecond());
                         if (CollectionUtils.isEmpty(schema.getEnum())) {
-                            schema.setEnum(getEnumValues(enumEl, schema.getType(), schema.getFormat(), context));
-                            addEnumExtensions(enumEl, schema, context);
+                            setEnumValues(enumEl, schema, context);
                         }
                     } else {
                         Schema<?> schemaWithSuperTypes = processSuperTypes(null, schemaName, type, definingElement, openAPI, mediaTypes, schemas, context, jsonViewClass);
@@ -601,9 +601,12 @@ public final class SchemaDefinitionUtils {
         return Pair.of(resultSchemaName, packageName + DOT + resultSchemaName);
     }
 
-    public static List<Object> getEnumValues(EnumElement type, String schemaType, String schemaFormat, VisitorContext context) {
+    public static void setEnumValues(EnumElement type, Schema schema, VisitorContext context) {
         var isProtobufGenerated = isProtobufGenerated(type);
         var enumValues = new ArrayList<>();
+
+        boolean cantDeserializeValues = false;
+
         for (EnumConstantElement element : type.elements()) {
 
             if (isProtobufGenerated) {
@@ -636,16 +639,32 @@ public final class SchemaDefinitionUtils {
             String jacksonValue = jsonPropertyAnn != null ? jsonPropertyAnn.stringValue(PROP_VALUE).orElse(null) : null;
             if (StringUtils.hasText(jacksonValue)) {
                 try {
-                    enumValues.add(ConvertUtils.normalizeValue(jacksonValue, schemaType, schemaFormat, context));
+                    enumValues.add(ConvertUtils.normalizeValue(jacksonValue, schema.getType(), schema.getFormat(), context));
                 } catch (JsonProcessingException e) {
                     warn("Error converting jacksonValue " + jacksonValue + " : to " + type + ":\n" + Utils.printStackTrace(e), context, element);
+                    if (!TYPE_STRING.equals(schema.getType())) {
+                        info("Changing enum type from " + type.getName() + " to " + TYPE_STRING, context, element);
+                    }
                     enumValues.add(element.getSimpleName());
+                    schema.setType(TYPE_STRING);
+                    schema.setFormat(null);
+                    cantDeserializeValues = true;
                 }
             } else {
+                if (!TYPE_STRING.equals(schema.getType())) {
+                    warn("Need to add @JsonProperty annotation for enum constant " + type.getName() + "." + element.getSimpleName(), context, element);
+                    info("Changing enum type from " + type.getName() + " to " + TYPE_STRING, context, element);
+                    schema.setType(TYPE_STRING);
+                    schema.setFormat(null);
+                    cantDeserializeValues = true;
+                }
                 enumValues.add(element.getSimpleName());
             }
         }
-        return !enumValues.isEmpty() ? enumValues : null;
+        if (!enumValues.isEmpty()) {
+            schema.setEnum(enumValues);
+        }
+        addEnumExtensions(type, schema, cantDeserializeValues, context);
     }
 
     /**

--- a/openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiEnumSpec.groovy
+++ b/openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiEnumSpec.groovy
@@ -1278,4 +1278,91 @@ class MyBean {}
         mapValueSchema.properties.value.format == 'int32'
         mapValueSchema.properties.prop.type == 'string'
     }
+
+    void "test enum with JsonValue and without JsonProperty"() {
+        when:
+        buildBeanDefinition('test.MyBean', '''
+package test;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import io.micronaut.http.annotation.Controller;
+import io.micronaut.http.annotation.Get;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@Controller("/hello")
+class HelloController {
+
+    @Get("/")
+    public COSEAlgorithmIdentifier index() {
+        return COSEAlgorithmIdentifier.EdDSA;
+    }
+}
+
+/**
+ * A number identifying a cryptographic algorithm. The algorithm identifiers SHOULD be values
+ * registered in the IANA COSE Algorithms registry, for instance, -7 for "ES256" and -257 for
+ * "RS256".
+ *
+ * @see <a
+ *     href="https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#typedefdef-cosealgorithmidentifier">§5.10.5.
+ *     Cryptographic Algorithm Identifier (typedef COSEAlgorithmIdentifier)</a>
+ */
+enum COSEAlgorithmIdentifier {
+  EdDSA(-8),
+  RS1(-65535);
+
+  private final long id;
+
+  @JsonValue
+  public long getId() {
+    return id;
+  }
+
+  COSEAlgorithmIdentifier(long id) {
+    this.id = id;
+  }
+
+  /**
+   * Attempt to parse an integer as a {@link COSEAlgorithmIdentifier}.
+   *
+   * @param id an integer equal to the {@link #getId() id} of a constant in {@link
+   *     COSEAlgorithmIdentifier}
+   * @return The {@link COSEAlgorithmIdentifier} instance whose {@link #getId() id} equals <code>id
+   *     </code>, if any.
+   * @see <a href="https://www.w3.org/TR/2021/REC-webauthn-2-20210408/#sctn-alg-identifier">§5.8.5.
+   *     Cryptographic Algorithm Identifier (typedef COSEAlgorithmIdentifier)</a>
+   */
+  public static Optional<COSEAlgorithmIdentifier> fromId(long id) {
+    return Stream.of(values()).filter(v -> v.id == id).findAny();
+  }
+
+  @JsonCreator
+  private static COSEAlgorithmIdentifier fromJson(long id) {
+    return fromId(id)
+        .orElseThrow(
+            () -> new IllegalArgumentException("Unknown COSE algorithm identifier: " + id));
+  }
+}
+
+@jakarta.inject.Singleton
+class MyBean {}
+''')
+        then: "the state is correct"
+        Utils.testReference != null
+
+        when: "The OpenAPI is retrieved"
+        OpenAPI openAPI = Utils.testReference
+        def enumSchema = openAPI.components.schemas.COSEAlgorithmIdentifier
+
+        then: "the components are valid"
+        enumSchema
+        enumSchema.type == "string"
+        !enumSchema.format
+        enumSchema.description == "A number identifying a cryptographic algorithm. The algorithm identifiers SHOULD be values registered in the IANA COSE Algorithms registry, for instance, -7 for \"ES256\" and -257 for \"RS256\"."
+        enumSchema.enum[0] == "EdDSA"
+        enumSchema.enum[1] == "RS1"
+    }
 }

